test/rules/non_constant_identifier_names_test.dart

```diff
@@ -10,9 +10,68 @@ main() {
   defineReflectiveSuite(() {
     defineReflectiveTests(NonConstantIdentifierNamesTest);
     defineReflectiveTests(NonConstantIdentifierNamesRecordsTest);
+    defineReflectiveTests(NonConstantIdentifierNamesPatternsTest);
   });
 }
 
+@reflectiveTest
+class NonConstantIdentifierNamesPatternsTest extends LintRuleTest {
+  @override
+  List<String> get experiments => ['patterns', 'records'];
+
+  @override
+  String get lintRule => 'non_constant_identifier_names';
+
+  @FailingTest(reason: 'Flow analysis fails w/ a Bad state exception')
+  test_patternForStatement() async {
+    await assertDiagnostics(r'''
+void f() {
+  for (var (ABC, def) = (0, 1); ABC <= 13; (ABC, def) = (def, ABC + def)) { }
+}
+''', [
+      lint(18, 3),
+    ]);
+  }
+
+  test_patternIfStatement() async {
+    await assertDiagnostics(r'''
+void f() {
+  if ([1,2] case [int ABC, int def]) print('hi');
+}
+''', [
+      lint(33, 3),
+    ]);
+  }
+
+  test_patternIfStatement_underscores() async {
+    await assertNoDiagnostics(r'''
+void f() {
+  if ([1,2] case [int _, int __]) print('hi');
+}
+''');
+  }
+
+  test_patternRecordField() async {
+    await assertDiagnostics(r'''
+void f() {
+  var (ABC, def) = (1, 2);
```

Comment:
Similarly here. Record patterns can have a single variable declaration, so this could be shortened to `var (A,) = (1,);`, and there's no ambiguity left as to which name is being flagged.